src/storage/src/storage.cc

```diff
@@ -1754,6 +1754,59 @@ Status Storage::Keys(const DataType& data_type, const std::string& pattern, std:
   return Status::OK();
 }
 
+Status Storage::Rename(const std::string& key, const std::string& newkey) {
+  const std::vector<std::string> keys = {key};
+  int64_t count = 0;
+  count = Exists(keys);
+  if (count <= 0) {
+    return Status::NotFound();
+  }
+
+  if (key != newkey) {
+    // zzl TODO: Is Status::Aborted() appropriate?
+    std::string value;
+    uint64_t ttl;
+    if (GetWithTTL(key, &value, &ttl) != Status::OK() || Del(keys) <= 0 || Set(newkey, value) != Status::OK()) {
+      return Status::Aborted();
+    }
+
+    if (ttl != -1) {
+      Expire(newkey, ttl);
+    }
+  }
+  return Status::OK();
+}
+
+Status Storage::Renamenx(const std::string& key, const std::string& newkey) {
+  const std::vector<std::string> keys = {key};
+  int64_t count = 0;
+  count = Exists(keys);
+  if (count <= 0) {
+    return Status::NotFound();
+  }
+
+  // in Renamenx, newkey should be not existed.
+  // otherwise, return Status::Corruption()
+  const std::vector<std::string> newkeys = {newkey};
+  // zzl TODO: Is Status::Corruption() appropriate?
+  count = Exists(newkeys);
+  if (count > 0) {
+    return Status::Corruption();
+  }
+
+  // zzl TODO: Is Status::Aborted() appropriate?
+  std::string value;
+  uint64_t ttl;
+  if (GetWithTTL(key, &value, &ttl) != Status::OK() || Del(keys) <= 0 || Set(newkey, value) != Status::OK()) {
```

Comment:
I'm not sure if calling other functions like GetWithTTL, Del, Set and Expire inside the rename(nx) function is allowed.